### ISSUE TITLE
Pin reservations are preserved when a composite device init fails

### ISSUE THREAD

**Issue by bennuttall** [gpiozero/gpiozero#804]
This won't affect most users, but will affect people hacking on gpiozero, testing things in the repl.

If you try to create a composite device, and it fails part way through, for example if it reaches a device which requires PWM, then the init will fail, your devices don't exist, but your pin reservations are preserved, meaning you can't fix the issue and recreate the object in the same session. Example:

```
ben@magicman:~/Projects/gpiozero/gpiozero (master) $ GPIOZERO_PIN_FACTORY=mock ipython

In [1]: from gpiozero import *                                                                       

In [2]: jh = JamHat()
...
PinPWMUnsupported:

In [3]: from gpiozero.pins.mock import MockPWMPin

In [4]: Device.pin_factory.pin_class = MockPWMPin

In [5]: jh = JamHat()
...
ValueError: pin 5 is already in use as a MockPin

In [6]: Device.pin_factory._reservations[5]                                                         
Out[6]: [<weakref at 0x7fe1dff3aae8; to 'LED' at 0x7fe1df787cf8>]
```

I used to get this a lot when hacking around with mock devices in the repl, before I discovered the `GPIOZERO_MOCK_PIN_CLASS` environment variable. So not a huge deal but it feels like pins should be released if the init failed.

Alternatively, could weakrefs be ignored or cleaned up?

**Comment by lurch** [comment]
Yeah, putting a try / except in the CompositeDevice `__init__` sounds like a good call.
